### PR TITLE
Initial PSK binding proposal.

### PR DESCRIPTION
**Pull request by dvorak42**
This adds the PSK binding via a custom PSK identity associated with extensions. Maybe a more general version would be useful to stick somewhere so that any extension which wants to do a binding doesn't need to hack it in?

Additional changes:
- Remove label swapping, since the PSK ClientHello binding should be sufficient to prevent reuse.
- Remove the AAD of KeyShareClientHello since the PSK ClientHello binding should be sufficient.
- Since now the encrypted_sni extension doesn't change at all as a result of HRR, simplify the wording to make sure the whole extension doesn't change. (server might not actually need to check if doesn't change, it just needs to be consistent in the processing of the extension, ignoring CH1 entirely, but saying it checks it doesn't change seems simplest.)

The naming/formatting of the Client Hello Binding section probably could use some work, not sure how to indicate that this is a hack onto an existing extension. (There was a nicer structure where there was a single PSK identity for extensions, and then we had a struct embedded in the PskBinderEntry, but it looks like the max PskBinderEntry size is 255 bytes, which is kind of limiting, though maybe doable if we think no more than 7 extensions will play these binding games.)

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -580,6 +570,35 @@ Section 9.2.) The client MUST NOT send a "cached_info" extension {{!RFC7924}}
 with a CachedObject entry whose CachedInformationType is "cert", since this
 indication would divulge the true server name.
 
+### Client Hello Binding {#client-hello-binding}
+
+To fully bind the ESNI extension contents to the rest of the ClientHello, an additional
+binding is inserted into the "pre_shared_key" extension. This is inserted as a special
+"ExtensionsPskIdentity" identity and "extensionsBinder" PskBinderEntry.
+
+       struct {
```
Comment (chris-wood) [inline draft-ietf-tls-esni.md:579]: nit: We probably need to wrap this in "~~~"

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -580,6 +570,35 @@ Section 9.2.) The client MUST NOT send a "cached_info" extension {{!RFC7924}}
...
+           opaque identity<7..2^16-1> = "CHExt" || extension_type;
```
Comment (chris-wood) [inline draft-ietf-tls-esni.md:580]: I wonder if folks will complain about possible identity collisions. The binders won't be the same since we use `encrypted_sni_binder_key` below, so perhaps it's not an issue?

**Review by chris-wood — APPROVED** [review]

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (dvorak42) [reply draft-ietf-tls-esni.md:580]: I'm hoping that we don't have accidental identity collisions (a real PSK with an identity of CHExtff07. An attacker could swap in their own PSK binder, but without knowing the encrypted_sni_binder_key, they shouldn't be able to generate a passing binder.

To have it written down somewhere, an alternative construction is having a single identity for all extensions, and then making the binder be a sequence of extension_type, binder pairs, but since the binders are limited to 255 bytes, that doesn't really extend very far. (Though I guess we could just stuff multiple extensions CHExtensionBinder1, CHExtensionBinder2, ..., but that seems hacky for a savings of a few bytes).
